### PR TITLE
Closes #1890 and #1900 to properly check for dates and fix other bugs in checkSpec

### LINKED ISSUE #1890: Bugfix: CheckSpec needs to check for Dates properly
## Expected Behavior

when `spec` has a `type: Date`, it should check to make sure that the column contains dates

## Current Behavior

the `mode()` function reads dates as numeric, so the test fails if `type: Date` is specified

## Possible Solution

use `class()` instead of `mode()` for dates

## Steps to Reproduce

1.
2.
3.
4.

### LINKED ISSUE #1900: Bugfix: allow for more complex specs to be checked properly
## Expected Behavior

users should be able to require `_all` columns of a given df, and also specify `type` for specific columns in specs and checkSpec() will check those columns appropriately, See below for example spec.

image

## Current Behavior

right now checkSpec() doesn't check those columns because `_all` is used for `required`, so it currently skips the type step as well due to some faulty logic.

## Possible Solution

updated logic in checkSpec to 

## Steps to Reproduce

1.
2.
3.
4.

### PR DESCRIPTION
**Pull request by lauramaxwell**
## Overview

- allows users to specify Dates in spec and checkspec will properly check the type
- changes from checking `mode()` to checking `class()` in checkSpec
  - most `numeric` types were converted to `integer` to pass this check.
- add testing for Date functionality

## Connected Issues

- Closes #1890
- Closes #1900

### CODE REVIEW

File: R/RunStep.R [hunk 1]
```diff
@@ -22,9 +22,9 @@
 #' 2. If a single parameter value is equal to "lData", the full lData object is passed to the function.
 #' 3. If a single parameter value is equal to "lSpec", the full lSpec object is passed to the function.
 #' 4. If a single parameter value is found in names(lMeta), that property is pulled from lMeta (e.g.
-#' lMeta${paramVal}) and passed to the function.
+#' lMeta$\{paramVal\}) and passed to the function.
```
Comment (zdz2101) [inline R/RunStep.R:25]: image

Is this the intentional "display" of this edit? Also there is a typo at line 15

File: R/util-checkSpec.R [hunk 2]
```diff
@@ -53,18 +53,18 @@ CheckSpec <- function(lData, lSpec) {
   # Check that all required columns in the spec are present in the data
   allCols <- c()
   missingCols <- c()
-  # remove all lSpec entries where we are requesting `_all` columns
-  lSpecDataFrames <- which(sapply(lSpec, function(x) names(x)[1] != "_all")) %>% names()
+  # remove all lSpec entries where we are requesting `_all` columns, and no
```
Comment (zdz2101) [inline R/util-checkSpec.R:56]:
```suggestion
  # remove all lSpec entries where we are requesting `_all` columns
```

File: tests/testthat/test-util-checkSpec.R [hunk 3]
```diff
@@ -218,3 +220,32 @@ test_that("proper message appears when all data frames require `_all` columns",
     " No required columns specified"
   )
 })
+
+test_that("All dates are handles correctly", {
```
Comment (zdz2101) [inline tests/testthat/test-util-checkSpec.R:224]: was this supposed to be a date based test?

**Review by zdz2101 — COMMENTED** [review]
Minor stuff, for the most part looks good

File: inst/workflow/4_modules/report_kri_country.yaml [hunk 4]
```diff
@@ -9,15 +9,15 @@ spec:
...
       required: true
     Numerator:
-      type: numeric
+      type: integer
     Denominator:
-      type: numeric
+      type: integer
     Metric:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_country.yaml:16]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_country.yaml [hunk 5]
```diff
@@ -9,15 +9,15 @@ spec:
...
     Score:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_country.yaml:18]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_site.yaml [hunk 6]
```diff
@@ -9,15 +9,15 @@ spec:
...
       required: true
     Numerator:
-      type: numeric
+      type: integer
     Denominator:
-      type: numeric
+      type: integer
     Metric:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_site.yaml:16]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_site.yaml [hunk 7]
```diff
@@ -9,15 +9,15 @@ spec:
...
     Score:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_site.yaml:18]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_site.yaml [hunk 8]
```diff
@@ -28,15 +28,15 @@ spec:
     _all:
       required: true
     Threshold:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_site.yaml:31]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_site.yaml [hunk 9]
```diff
@@ -28,15 +28,15 @@ spec:
...
     Numerator:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_site.yaml:33]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_site.yaml [hunk 10]
```diff
@@ -28,15 +28,15 @@ spec:
...
     Denominator:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_site.yaml:35]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_site.yaml [hunk 11]
```diff
@@ -28,15 +28,15 @@ spec:
...
     LogDenominator:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_site.yaml:37]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_site.yaml [hunk 12]
```diff
@@ -28,15 +28,15 @@ spec:
...
     Metric:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_site.yaml:39]:
```suggestion
      type: numeric
```

File: tests/testthat/test-util-checkSpec.R (hunk 3 above)
Comment (lauramaxwell) [reply tests/testthat/test-util-checkSpec.R:224]: oh whoops- i started writing a whole new test, but then ended up just testing dates in the "type" test above by including the  `SnapshotDate` column. I'll delete this now.

File: inst/workflow/3_reporting/Bounds.yaml
Comment (samussiah) [inline inst/workflow/3_reporting/Bounds.yaml:1]: `Numerator`, `Denominator`, and `Flag` can be switched to `integer` here.

File: R/RunStep.R (hunk 1 above)
Comment (lauramaxwell) [reply R/RunStep.R:25]: yes, this is the intended display. i'll edit the typo up top. thanks for catching that!

File: inst/workflow/4_modules/report_kri_country.yaml [hunk 13]
```diff
@@ -28,15 +28,15 @@ spec:
     _all:
       required: true
     Threshold:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_country.yaml:31]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_country.yaml [hunk 14]
```diff
@@ -28,15 +28,15 @@ spec:
...
     Numerator:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_country.yaml:33]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_country.yaml [hunk 15]
```diff
@@ -28,15 +28,15 @@ spec:
...
     Denominator:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_country.yaml:35]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_country.yaml [hunk 16]
```diff
@@ -28,15 +28,15 @@ spec:
...
     LogDenominator:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_country.yaml:37]:
```suggestion
      type: numeric
```

File: inst/workflow/4_modules/report_kri_country.yaml [hunk 17]
```diff
@@ -28,15 +28,15 @@ spec:
...
     Metric:
-      type: numeric
+      type: integer
```
Comment (samussiah) [inline inst/workflow/4_modules/report_kri_country.yaml:39]:
```suggestion
      type: numeric
```

**Review by samussiah — APPROVED** [review]
Logic makes sense to me!  For the reporting data layer `Results$Metric`, `Results$Score`, and everything in `Bounds` should remain `numeric`.

**Pull request merged**